agda: 2.6.2.2 -> 2.6.3

iblech opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

[Agda 2.6.3 is out!](https://hackage.haskell.org/package/Agda-2.6.3/changelog) This pull request brings the new version to nixpkgs.

This pull request does not contain the required changes to automatically generated files for our Haskell package set, as I believe to be best practice in nixpkgs. To test this pull request locally, you hence first need to manually update these:
```
./maintainers/scripts/haskell/update-stackage.sh
./maintainers/scripts/haskell/update-hackage.sh
./maintainers/scripts/haskell/regenerate-hackage-packages.sh
nix-shell -I nixpkgs=$PWD -p "agda.withPackages (p: [ p.agda-categories p.standard-library p.cubical p.agdarsec p.agda-prelude p.functional-linear-algebra ])"
```

You can also test it out live on the [Agdapad](https://agdapad.quasicoherent.io/).

Because of minor incompatibilities between Agda 2.6.2.2 and Agda 2.6.3, this pull request also includes the following changes:

- Update of the stdlib from 1.7.1 to 1.7.2-rc1
- Update of cubical to current HEAD and including a soon-to-be-merged pull request to cubical
- Update of agda-prelude to current HEAD
- Setting `LC_ALL = "C.UTF-8"` for all Agda package builds, as per #172752

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

iblech committed:
agda: 2.6.2.2 -> 2.6.3

iblech commented:
Awesome, thank you @sternenseemann for tending to this PR. :-)

In the force-push I upgraded stdlib from 1.7.2-rc1 to the now-released 1.7.2, and cubical-pre-whatever to the now-current HEAD, thereby no longer requiring to pull in a then-not-yet-merged cubical PR.

ncfavier commented on pkgs/build-support/agda/default.nix:
```diff
@@ -81,6 +81,9 @@ let
           runHook postInstall
         '';
 
+        # see issue #172752
```
```suggestion
        # see https://github.com/NixOS/nixpkgs/issues/172752
```

ncfavier commented on pkgs/development/libraries/agda/agda-prelude/default.nix:
```diff
@@ -1,14 +1,14 @@
-    rev = version;
-    sha256 = "0j2nip5fbn61fpkm3qz4dlazl4mzdv7qlgw9zm15bkcvaila0h14";
+    rev = "3d143d6d0a3f75966602480665623e87233ff93e";
+    sha256 = "sha256-ILhXDq788vrceMp5mCiQUMrJxeLPtS4yGtvMHMYxzg8=";
```
```suggestion
    hash = "sha256-ILhXDq788vrceMp5mCiQUMrJxeLPtS4yGtvMHMYxzg8=";
```

ncfavier commented on pkgs/development/libraries/agda/standard-library/default.nix:
```diff
@@ -2,13 +2,13 @@
     owner = "agda";
     rev = "v${version}";
-    sha256 = "0khl12jvknsvjsq3l5cbp2b5qlw983qbymi1dcgfz9z0b92si3r0";
+    sha256 = "sha256-vvbyfC5+Yyx18IDikSbAAcTHHtU6krlz45Fd2YlwsBg=";
```
```suggestion
    hash = "sha256-vvbyfC5+Yyx18IDikSbAAcTHHtU6krlz45Fd2YlwsBg=";
```

iblech replied:
Thank you for this and your other review comments, very much appreciated. :-)

turion commented:
Thanks a lot!

As described in https://nixos.org/manual/nixpkgs/stable/#maintaining-the-agda-package-set-on-nixpkgs, It's possible to check reverse dependencies of updated packages, or even the whole package set:

@ofborg build agda.passthru.tests.allPackages

> This pull request does not contain the required changes to automatically generated files for our Haskell package set, as I believe to be best practice in nixpkgs.

Does that mean that this PR should be targeted against `haskell-updates`?

CC @alexarice

ncfavier commented:
Yes, we'll likely want to rebase this on `haskell-updates` once https://github.com/NixOS/nixpkgs/pull/213089 is merged and a new update cycle starts.

alexarice commented:
Won't agda update automatically when haskell updates?

ncfavier commented:
It's pinned to 2.6.2.2 by the Stackage LTS, so we'll still need [this override](https://github.com/NixOS/nixpkgs/pull/213896/files#diff-06cdf278a2cf7f0e7793a4a8d19cbdd6e6a3a3722482708833ef9270b65d7d56R2337-R2338)

maralorn commented:
This is an example of a maintained package where the maintainer wants to maintain a version newer than in stackage.

For this reason, I vote to replace the [override in configuration-common.nix](https://github.com/NixOS/nixpkgs/pull/213896/files#diff-06cdf278a2cf7f0e7793a4a8d19cbdd6e6a3a3722482708833ef9270b65d7d56R2337-R2338) with a _commented_ entry in [the stackage-update script.](https://github.com/NixOS/nixpkgs/blob/09d38c4d143ee3c338ce988a8bf1ff5e996809e8/maintainers/scripts/haskell/update-stackage.sh#L54)

That way, we don’t need to bump the override every time when Agda receives an update. I also vote that we should do this for other similar cases.

@NixOS/haskell do you agree?

iblech commented:
Thank you for your input @maralorn :-)

I'm not listed as an Agda maintainer but definitely care about this package a lot. Your suggestion sounds very reasonable to me, though I wouldn't claim to be exactly familiar about the nixpkgs Haskell ecosystem and hence my judgment shouldn't carry much weight.

In nixpkgs, we strive to always have a compatible set of Agda the compiler and the various Agda library in existence. It is conceivable that a new Agda version gets released which is incompatible with the latest released standard library. In this case we might need to hold Agda back. However, I expect this occurrence to be very rare; should it occur, it could be dealt with an exception. So to change the default to "always follow Hackage" seems good to me!

turion commented:
I think it's great to send this agda update ahead, so we can fix the agda package set (which is often broken by the update). But I wouldn't want to merge it before agda on hackage/stackage/haskellPackages has updated in order to avoid version overrides, so I'd target this PR against `haskell-updates`, and merge it when the bump comes anyways. Is that a viable workflow as well?

maralorn commented:
Well, we need to be clear here.
Currently, the hackage snapshot in nixpkgs is so old, that Agda 2.6.3 does not exist. That is why this PR needs to wait for the next bump on haskell-updates.
But even then, Stackage LTS will be pinning our default version of Agda to 2.6.*.
Waiting until Agda reaches 2.6.3 all by itself will therefore probably still take months.
The purpose of this PR would be to jump that gun.

maralorn commented:
Oh, wait, this is a minor version update. So it might actually be faster than months, sorry. But as of now stackage LTS hasn’t bumped agda.

turion commented:
I don't know whether there is any advantage to merging this now already. If people are very eager to have agda 2.6.3, then yes. (It seems @iblech is.) But we have to remember to remove the version override later.

ncfavier commented:
We cannot merge this now because it assumes that `haskellPackages.Agda_2_6_3` exists, which it does not on master. We need to wait for the next Hackage bump and then re-base this PR on `haskell-updates`, as discussed.

Drafting to avoid further confusion.

ncfavier commented:
@iblech you can now rebase against `haskell-updates`, targetting https://github.com/NixOS/nixpkgs/pull/215925.

ncfavier commented on pkgs/development/libraries/agda/agda-categories/default.nix:
```diff
@@ -12,10 +12,10 @@ mkDerivation rec {
   postPatch = ''
     substituteInPlace agda-categories.agda-lib \
-      --replace 'standard-library-1.7' 'standard-library-1.7.1'
+      --replace 'standard-library-1.7' 'standard-library-1.7.2'
```
Agda's [version matching system](https://agda.readthedocs.io/en/latest/tools/package-system.html#version-numbers) is somewhat insane: there's no middle ground between "no version constraint" and "exact match down to patch". In order to avoid having to change this with every standard-library update I think we should drop the constraint altogether:

```suggestion
      --replace 'standard-library-1.7' 'standard-library'
```

But beware that the left hand side changes in later versions of agda-categories, so we really want something like

```sh
sed -Ei 's/standard-library-[0-9.]+/standard-library/' agda-categories.agda-lib
```

Maybe even convince upstream to drop the constraint.

Updating agda-categories will take some work anyway, from what I can tell (Everything.agda has to be generated, which seems to involve git).

iblech replied:
Good idea! In the imminent rebase and force-push, I include your sed suggestion.

I am not entirely sure what you mean by the comment on Everything.agda. Currently we don't do anything ourselves regarding Everything.agda and everything seems fine :-)

ncfavier replied:
Right, but future updates will not be fine. See https://github.com/agda/agda-categories/commit/8643687021b321420665bdf7f9a27983b564291b

iblech replied:
Ah, I see! Thank you for pointing this out.

In the next force-push, I will upgrade to categories v0.1.7.1a (thereby exhibiting the issue you raised) and fix it by generating Everything.agda ourselves.

The command listed in the Makefile doesn't seem suitable for us, because we are not working in a Git checkout. But the following command, added to postPatch, seems to do its job just fine:

```
find src -name '*.agda' | sed -e 's|^src/[/]*|import |' -e 's|/|.|g' -e 's/.agda//' -e '/import Everything/d' | LC_COLLATE='C' sort > Everything.agda
```

Compilation is still running, I will report once it finishes.

ncfavier reviewed (approved):
I was able to build `agda.passthru.tests.allPackages` after rebasing on `haskell-updates`, so LGTM aside from the agda-categories comment.

As another aside it would be nice if `agda.tests` and `agda.passthru.tests` were the same thing... EDIT: https://github.com/NixOS/nixpkgs/pull/215990

iblech commented:
For those following along, locally I rebased to haskell-updates and implemented @maralorn's suggestion. I will then also bump cubical to the current HEAD, test everything and then force-push. I expect this to take approximately one hour.

iblech commented:
Did so. As this pull request touches `update-stackage.sh`, here are the steps required to test this locally:

0. Check out the branch.
1. Edit maintainers/scripts/haskell/update-stackage.sh so the script doesn't abort if no new Stackage version exists.
2. Run `./maintainers/scripts/haskell/update-stackage.sh`
3. Run `./maintainers/scripts/haskell/update-hackage.sh`
4. Run `./maintainers/scripts/haskell/regenerate-hackage-packages.sh`
5. Run `nix-shell -I nixpkgs=$PWD -p "agda.withPackages (p: [ p.agda-categories p.standard-library p.cubical p.agdarsec p.agda-prelude p.functional-linear-algebra ])"`

Only then will Agda-2.6.3 be visible.

ncfavier commented:
You don't need step 3, actually. Could you make a commit with the results of steps 2 and 4, so that this PR actually updates Agda?

As a nice side-effect of using `Agda` rather than `Agda_2_6_3`, we get the [`-foptimise-heavily`](https://github.com/NixOS/nixpkgs/blob/d92f1f04cd14ba2217f7184a8da724a5a2a13086/pkgs/development/haskell-modules/configuration-nix.nix#L899-L902) override, which only applies to `Agda` (ideally such flags would be set at the level of hackage2nix so that they apply to all versions...)

iblech commented:
Thank you for your feedback, @ncfavier! I was under the impression that the regeneration should only be done once, by the Haskell team, not in random pull requests. But sure, I will incorporate your suggestion in the next force-push :-)

iblech commented:
Did so. You can now test this pull request locally with a simple:

```
nix-shell -I nixpkgs=https://github.com/iblech/nixpkgs/archive/patch-agda-2.6.3.tar.gz -p "agda.withPackages (p: [ p.agda-categories p.standard-library p.cubical p.agdarsec p.agda-prelude p.functional-linear-algebra ])"
```

sternenseemann commented:
@ofborg test agda

sternenseemann commented:
Should we add you as a maintainer for agda, @iblech?

sternenseemann commented on pkgs/build-support/agda/default.nix:
```diff
@@ -81,6 +81,9 @@ let
         '';
 
+        # see https://github.com/NixOS/nixpkgs/issues/172752
+        LC_ALL = "C.UTF-8";
```
This locale doesn't exist on darwin unfortunately, you may want to add a conditional to have it be `en_US.UTF-8` on that platform or something – if darwin support is imporant to you, that is.

(It's a real shame and major annoyance that there does not seem to be a _standard_ UTF-8 supporting locale on darwin, unless someone can enlighten me.)

ncfavier replied:
The `allPackages` build succeeds on x86_64-darwin, though: https://logs.nix.ci/?key=nixos/nixpkgs.213896&attempt_id=b2775f53-d698-4681-85f3-0d1ecd918022

sternenseemann replied:
Okay, then the linked issue doesn't affect darwin apparently, as the locale indeed doesn't work:

```
sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8): No such file or directory
```

Maybe we can just not set it for darwin then…

iblech replied:
Okay, will try that!

sternenseemann commented on pkgs/development/libraries/agda/cubical/default.nix:
```diff
@@ -2,17 +2,15 @@
 mkDerivation rec {
   pname = "cubical";
-  version = "0.4";
+  version = "0.5pre6b1ce0";
```
Why are you not using unstable here, but in the other package(s)?

iblech replied:
No particular reason, the next force-push will harmonize the version strings :-)

iblech commented:
Awesome, happy to see this PR merged. I really enjoyed the process from the first draft to the final commit. Thank you all for your comments. :-) On to Agda 2.6.4 :-)

sternenseemann commented:
Fallout from this seems to be bladga – not sure if anyone cares though:

- https://hydra.nixos.org/build/209371026/nixlog/1
- https://github.com/isovector/blagda/issues/2
